Added sequence generation option by giving a generator method

btezergil opened the pull request:
# Proposed Changes
- Added :generator option to evolve method to use sequence generator function
- Converted build-population into a multi-arity method that supports sequence generator function

## Pre-merge Checklist

- [ ] Write + run tests
- [ ] Update CHANGELOG and increment version
- [ ] Update README and relevant documentation

btezergil-atlassian committed:
Added sequence generation option by giving a generator method

btezergil-atlassian committed:
Added new project qualifier to push my version to clojars

nnichols commented on src/nature/core.cljc:
```diff
@@ -26,8 +26,11 @@
-         monitors (:monitors options)]
-     (loop [population (io/build-population population-size allele-set genome-length fitness-function)
+         monitors (:monitors options)
+         sequence-generator-function (:generator options)]
```
Could you add this to the docstring for this function?

btezergil replied:
The new commit addresses this, you can find my detailed response below.

nnichols commented on src/nature/initialization_operators.cljc:
```diff
@@ -32,6 +32,10 @@
-  "Build `population-size` individuals by invoking `build-individual` on random, conforming genetic sequences."
-  [population-size alleles sequence-length fitness-function]
-  (repeatedly population-size #(build-individual (generate-sequence alleles sequence-length) fitness-function)))
+  "Build `population-size` individuals by invoking `build-individual` on random, conforming genetic sequences.
```
This looks good, but I'd really appreciate a test with a trivial sequence-generator-function if you wouldn't mind

btezergil-atlassian committed:
Refactored the evolve method, extracted the common part into a function
and removed the unused function parameters for sequence generation
functon case

btezergil commented:
I've made updates to core.cljc in my latest commit. I think the initial implementation lacked some clarity in the sense that it made two function parameters (`allele-set` and `genome-length`) obsolete. Therefore, I've extracted the common parts into a function with a docstring, and divided the evolve method into two parts, the new one only gets a generator function instead of an allele set and genome length. I've also added an explanatory docstring into the new method.

While the new changes were not necessary in terms of functionality, I think eliminating the unused parameters made the overall code more readable. If you are okay with this version, I'll be adding the necessary tests to finalize the development related to this. 

There is a middle commit that is unrelated, I'll be removing that part and squashing the commits together after you are okay with all changes. 

Also, as a final point, I've realized after a couple of hours that the mutation and crossover functions require nature's individual structure to be returned, we can make this clearer in docstrings. WDYT?

btezergil-atlassian committed:
Added insert-new option to insert new individuals in every cycle

btezergil-atlassian committed:
Changed prints into logs

btezergil-atlassian committed:
Parallelization changes

nnichols commented:
This project is a few years old, and I don't actively maintain it. I also wouldn't be able to merge this PR with the Maven artifact changes and without tests.

You are free to fork this repository and maintain the code you'd like, but I will be archiving this repository in the near future.
